### PR TITLE
D2M: Matmul Multicore blocking, additional fixes

### PR DESCRIPTION
**Pull request by jdesousa-TT**
### Problem description
We would like to achieve parity in main with the matmul and reductions we've achieved on branch.

### What's changed
Separate things addressed in this PR:
1. ~~Override matmul k-blocking shapes to default to k_block_size=1. This allows us to fit large matmuls for now, until we have a smart way to decide this block factor. _(if we don't want this in main, we can limit the size of our tests. open to feedback here.)~~
2. Add a `matmul-block-factors` command line option to override the block factors used by OptimizeTensorLayout on matmuls. This is an interim solution which allows us to test different configs of matmul. This is required for the follow-up PR which changes the meaning of blockFactors to be based on L1 availability. We will not have an intelligent way to select this factor for the present time.
3. Replace ToLayout on the result of generic ops with ViewLayout. When we block outputs, we don't actually produce output blocks on different cores than originally intended. ToLayout will fail here because it tries to move data from cores that were not included in the compute. This unblocks matmul and reductions (cc: @phizalev-TT). This is a side effect of using grid shape to encode blocking information. 
4. Fix `blockedView` to not multiply `dim` by the blockFactor if the dimension is a reduction dim. This is because we need to gather all of the `k` block from all cores to mcast or reduce.
5. Override the shape that we intiialize cbs on to the device worker grid shape at all times. **I'm open to suggestions on better ways to do this.** This becomes a problem because we use grid shape to encode blocking information, which again can sometimes be larger than the actual worker grid. There is no side effect to initializing cbs on cores we will not do work on, but if you pass coordinates outside of the worker grid range, it will fail.
6. Added multicore matmul tests that we used to benchmark against ttnn. These are the limit of what we can currently accomplish. 

### Checklist
- [X] New/Existing tests provide coverage for changes

### CODE REVIEW

**Review by xanderchin — APPROVED** [review]
If we don't have one already, create an issue for intelligently selecting matmul k block factor.

**Comment by jdesousa-TT** [comment]
Just a heads up on this one, the builder tests are failing awaiting the disabling of `fp32DstAccum` mode which is in flight.

**Review by bgrady-tt — APPROVED** [review]
LGTM 👍

File: lib/Dialect/TTIR/Transforms/OptimizeTensorLayout.cpp [hunk 1]
```diff
@@ -147,6 +147,21 @@ calculateOptimalBlockFactors(ArrayRef<AffineMap> indexingMaps,
   return inverse.compose(flattenedBlockFactors);
 }
 
+// This override is the "trivial" way to k-block matmuls so that we can test and
+// benchmark large matmuls. We set the block factor for k to the size of the k
+// shard dimension. This means each block is one tile. In the future, we should
+// intelligently pick this factor to optimize memory usage.
+static void
+overrideMatmulKBlockFactor(OpBuilder &rewriter, Value in0Tensor,
+                           SmallVector<int64_t> &blockFactors,
```
Comment (vroubtsovTT) [inline lib/Dialect/TTIR/Transforms/OptimizeTensorLayout.cpp]: Passing `blockFactors` by non-const reference here looks ugly in the middle of the signature. It is also not asserted that it actually has size > 2. Instead of using an out-parameter here, is it possible to instead return a value and use it to set whichever `blockFactors` slot by the caller?

**Review by vroubtsovTT — APPROVED** [review]
I think it's hard to understand what OptimizeToLayout does these days, but looks ok as WIP.

**Review by vroubtsovTT — COMMENTED** [review]
> Replace ToLayout on the result of generic ops with ViewLayout. When we block outputs, we don't actually produce output blocks on different cores than originally intended. ToLayout will fail here because it tries to move data from cores that were not included in the compute. 

Actually, I am now second guessing this motivation. ToLayout was added to ttir.generic input/output for a reason, that being ToLayout it responsible for converting b/w what ttir.generic might choose to use internally and what the surrounding IR expects, both on its inputs and its output. That was my interpretation of #2762.

A sequence of two generics would be something like
```mlir
%0 = ...

%in_A = to_layout %0 ...
%out_A = generic %in_A, ...
%1 = to_layout %out_A ...

%in_B = to_layout %1 ...
%out_B = generic %in_B, ...
%2 = to_layout %out_B ...
```
For a sequence of generics like that, the pair of to_layouts `%1 = to_layout %out_A ... ; %in_B = to_layout %1 ...` will be collapsed to a single "last one wins" to_layout `%in_B = to_layout %out_A ...` by the canonicalizer. And if the resulting from-layout and to-layout are actually identical and the layout conversion is actually a no-op, that single to_layout will be removed by the folder.

But if the layout change is not a no-op, to_layout must remain and has the responsibility of actually taking care of the change, including core grid changes and whatever data movement is necessary. view_layout cannot do this... am I missing something?

File: test/python/golden/test_metal_matmul.py [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+import pytest
+import torch
+from typing import List
+
+from ttir_builder.utils import compile_to_flatbuffer
+from ttir_builder import Operand, TTIRBuilder, Shape
+from ttmlir.ir import *
+
+
+@pytest.mark.fails_golden
```
Comment (nsmithtt) [inline test/python/golden/test_metal_matmul.py:14]: Do we know why?

**Review by phizalev-TT — APPROVED** [review]

**Comment by jdesousa-TT** [comment]
> If we don't have one already, create an issue for intelligently selecting matmul k block factor.

@xanderchin #4075

File: lib/Dialect/TTIR/Transforms/OptimizeTensorLayout.cpp (hunk 1 above)
Comment (jdesousa-TT) [reply lib/Dialect/TTIR/Transforms/OptimizeTensorLayout.cpp]: I'm going to change this override to be a configurable flag called `--matmul-block-factors` as an interim solution for testing until we have an intelligent solution (#4075). 

The motivation is that the loop tiling PR which will come soon after this one actually doesn't want to override this way, and the default of all 1s actually makes sense.

File: test/python/golden/test_metal_matmul.py (hunk 2 above)
Comment (jdesousa-TT) [reply test/python/golden/test_metal_matmul.py:14]: I don't think we do know why. It might just be that we lose too much accuracy in fp32 without `fp32DstAccum` turned on. 

@wenbinlyuTT, for the SFPU ops you were bringing up, did the allclose checks pass when you had `fp32DstAccum` enabled?

File: lib/Dialect/TTIR/Transforms/OptimizeTensorLayout.cpp [hunk 3]
```diff
@@ -151,9 +151,11 @@ namespace {
 struct TTIRGenericTensorLayoutRewriter : public OpRewritePattern<GenericOp> {
   TTIRGenericTensorLayoutRewriter(MLIRContext *context,
                                   SmallVector<int64_t> workerGridShape,
-                                  unsigned dstRegisterSizeTiles)
+                                  unsigned dstRegisterSizeTiles,
+                                  SmallVector<int64_t> matmulBlockFactors)
       : OpRewritePattern<GenericOp>(context), workerGridShape(workerGridShape),
-        dstRegisterSizeTiles(dstRegisterSizeTiles) {}
+        dstRegisterSizeTiles(dstRegisterSizeTiles),
```
Comment (nsmithtt) [inline lib/Dialect/TTIR/Transforms/OptimizeTensorLayout.cpp:157]: Shouldn't we get rid of dstRegisterSize tiles here?  I was thinking that we'd largely revert my commit that introduced the use of dstRegisterSizeTiles here.

File: lib/Dialect/TTIR/Transforms/OptimizeTensorLayout.cpp [hunk 4]
```diff
@@ -170,6 +172,20 @@ struct TTIRGenericTensorLayoutRewriter : public OpRewritePattern<GenericOp> {
...
+        op.getIteratorTypesValue()[2] == ttcore::IteratorType::Reduction &&
+        !matmulBlockFactors.empty()) {
+      assert(matmulBlockFactors.size() == 3 &&
+             "Matmul block factors must be of size 3");
+      for (auto [i, factor] : llvm::enumerate(matmulBlockFactors)) {
+        if (factor != 0) {
+          blockFactors[i] = factor;
+        }
+      }
+    }
```
Comment (nsmithtt) [inline lib/Dialect/TTIR/Transforms/OptimizeTensorLayout.cpp:188]: Let's break this into a helper func

File: lib/Dialect/TTIR/Transforms/OptimizeTensorLayout.cpp (hunk 3 above)
Comment (jdesousa-TT) [reply lib/Dialect/TTIR/Transforms/OptimizeTensorLayout.cpp:157]: It's removed in the tiling PR #4070. In this PR the blocking here still serves its original purpose of respecting dst sizing.
